### PR TITLE
Preview environment for every pull request using Uffizzi 

### PR DESCRIPTION
**Pull request by waveywaves**
## Hey, I just made a Pull Request!

- add new github actions workflow to create new backstage app, build it and create a new preview using uffizzi. uffizzi-build.yml builds the image from the PR. Once it completes, uffizzi-preview.yml is triggered in the context of the default branch to create the preview and post a comment to the Pull request
- add .github/uffizzi folder with the docker-compose template and app config for backstage created by uffizzi.

I am creating this pull request after improving the product and ensuring the stability through other Open Source Integrations done which can be seen [here](https://uffizzi.notion.site). Also have implemented the feedback from last time working on the Uffizzi integration with Backstage https://github.com/backstage/backstage/pull/13900

The PoC for this PR can be seen here https://github.com/waveywaves/backstage/pull/17
A comment will be posted on every PR created like [this](https://github.com/waveywaves/backstage/pull/17#issuecomment-1341646619) 
And the preview built off of the PoC PR can be seen [here](https://app.uffizzi.com/github.com/waveywaves/backstage/pull/17)

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by waveywaves** [comment]
cc @jhaals @benjdlambert @freben

File: .github/workflows/uffizzi-build.yml [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        run: |
+          yarn build:all
+        working-directory: ./example-app
+
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v2
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
```
Comment (Rugvip) [inline .github/workflows/uffizzi-build.yml:84]: This secret won't be accessible by forks right? That likely means these previews wouldn't be available in most PRs. Is there some other way we can go about managing or triggering this? Perhaps move the build step down to the deploy run?

**Review by Rugvip — COMMENTED** [review]
Thank you for taking another stab at this! 👍 

Drive-by review because I spotted a thing

File: .github/workflows/uffizzi-build.yml (hunk 1 above)
Comment (waveywaves) [reply .github/workflows/uffizzi-build.yml:84]: There is no need for this. The image will be pushed to the uffizzi registry. Overlooked this bit of the github workflow. Removing it now.

Comment (waveywaves) [reply .github/workflows/uffizzi-build.yml]: @Rugvip updated the PR

**Review by djamaile — COMMENTED** [review]
Hi! CI is currently failing because of prettier. `yarn prettier --write .github` should work!

**Review by djamaile — COMMENTED** [review]
Hmm not sure where the tsc errors come from. Can you rebase from master and see if they persist?

**Comment by waveywaves** [comment]
> Hmm not sure where the tsc errors come from. Can you rebase from master and see if they persist?

@djamaile Done. Let's see the results.

**Comment by waveywaves** [comment]
@djamaile :crossed_fingers: rebased :crossed_fingers:

**Comment by waveywaves** [comment]
@djamaile the CI is green now ! :heavy_check_mark:

File: .github/workflows/uffizzi-build.yml [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+      - name: backstage create-app | delete if already exists
+        run: |
+          rm -rf ./example-app
+          npx @backstage/create-app
+        env:
+          BACKSTAGE_APP_NAME: example-app
+
+      - name: yarn build
+        run: |
+          yarn build:all
```
[the remark this answers is not in the document]
Comment (waveywaves) [reply .github/workflows/uffizzi-build.yml:77]: Added tsc

**Review by djamaile — COMMENTED** [review]
To me it looks good now. But it is hard too know for sure if we can't test it before we merge. Is there a way we can do that? @freben

**Review by djamaile — APPROVED** [review]

File: .github/uffizzi/uffizzi.production.app-config.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+app:
+  title: Backstage Uffizzi Environment
+  baseUrl: ${UFFIZZI_URL}
+
+backend:
+  baseUrl: ${UFFIZZI_URL}
+  cors:
+    origin: ${UFFIZZI_URL}
+    methods: [GET, POST, PUT, DELETE]
+    credentials: true
```
Comment (freben) [inline .github/uffizzi/uffizzi.production.app-config.yaml:10]: This one doesn't have any catalog data. In a future iteration (not necessary for this first PR) we should probably make it have some of that locally in the image at least (maybe the example data that's already in `packages/catalog-model/examples`?), so we can get more of a rich testing experience.

**Review by freben — COMMENTED** [review]
OK some more comments!

File: .github/workflows/uffizzi-build.yml [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+          restore-keys: |
+            ${{ runner.os }}-yarn-
+      - name: yarn install
+        if: steps.cache-modules.outputs.cache-hit != 'true'
+        run: yarn install --immutable
+      # End of yarn setup
+
+      - name: backstage create-app | delete if already exists
+        run: |
+          rm -rf ./example-app
```
[the remark this answers is not in the document]
Comment (waveywaves) [reply .github/workflows/uffizzi-build.yml:69]: Added this step as the github action workflow preserves the `example-app` directory. When running `backstage create-app` again, we get an error saying that the directory already exists.

File: .github/workflows/uffizzi-build.yml [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+      - name: yarn build
+        run: |
+          tsc
+          yarn build:all
+        working-directory: ./example-app
+
+      - name: Use Uffizzi's backstage app config
+        run: |
+          echo "------------------------ COPY UFFIZZI APP CONFIG"
```
[the remark this answers is not in the document]
Comment (waveywaves) [reply .github/workflows/uffizzi-build.yml:82]: removing it

File: .github/uffizzi/uffizzi.production.app-config.yaml (hunk 3 above)
Comment (waveywaves) [reply .github/uffizzi/uffizzi.production.app-config.yaml:10]: I will keep this for now and update this change in a subsequent PR.

**Review by freben — APPROVED** [review]
Let's try!

**Comment by waveywaves** [comment]
@djamaile the change you made was to the example-app which already exists in the repo. The workflow creates a new example-app using the cli.

**Pull request merged**